mappings/doab.py

```diff
@@ -1,130 +1,178 @@
-                        './datacite:date/text()',
-                        './datacite:date/@type'
-                    ],
-                    '{0}|{1}'
-                )
-            ],
-            'languages': [('./dc:language/text()', '||{0}')],
-            'extent': ('./oapen:pages/text()', '{0} pages'),
-            'abstract': ('./dc:description/text()', '{0}'),
-            'subjects': [('./datacite:subject/text()', '{0}||')],
-            'has_part': [(
-                './dc:identifier/text()',
-                '1|{0}|doab|text/html|{{"reader": false, "download": false, "catalog": false, "embed": true}}'
-            )],
-            'rights': [
-                (
-                    [
-                        './oaire:licenseCondition/@uri',
-                        './oaire:licenseCondition/text()'
-                    ],
-                    'doab|{0}||{1}|'
-                ),
-                (
-                    [
-                        './datacite:rights/@uri',
-                        './datacite:rights/text()'
-                    ],
-                    'doab|{0}||{1}|'
-                )
-            ]
-        }
+    def map_doab_record(self, doab_record, namespaces) -> Record:
+        identifiers, source_id = self._get_identifers(doab_record, namespaces=namespaces)
+        title = doab_record.xpath('./dc:title/text()', namespaces=namespaces) + doab_record.xpath('./datacite:creator/text()', namespaces=namespaces)
+
+        if identifiers is None or len(identifiers) == 0 or title is None or len(title) == 0:
+            return None
+
+        relations = doab_record.xpath('./dc:relation/text()', namespaces=namespaces)
+        publishers = doab_record.xpath('./dc:publisher/text()', namespaces=namespaces)
+        languages = doab_record.xpath('./dc:language/text()', namespaces=namespaces)
+        extent = doab_record.xpath('./oapen:pages/text()', namespaces=namespaces)
+        abstract = doab_record.xpath('./dc:description/text()', namespaces=namespaces)
+        
+        return Record(
+            uuid=uuid4(),
+            frbr_status=FRBRStatus.TODO.value,
+            cluster_status=False,
+            source=Source.DOAB.value,
+            source_id=source_id,
+            identifiers= identifiers,
+            authors=self._get_authors(doab_record, namespaces=namespaces),
+            contributors=self._get_contributors(doab_record, namespaces=namespaces),
+            title=title[0],
+            is_part_of=[f'{part}||series' for part in relations if part],
+            publisher=[f'{publisher}||' for publisher in publishers if publisher],
+            spatial=doab_record.xpath('./oapen:placepublication/text()', namespaces=namespaces),
+            dates=self._get_dates(doab_record, namespaces=namespaces),
+            languages=[f'||{language}' for language in languages if language],
+            extent=f'{extent[0]} pages' if extent else None,
+            abstract=abstract[0] if abstract else None,
```

Comment:
Nice - i like this - especially because it'll check if the array is both not none and not empty